data-prepper-plugins/dynamodb-source/src/main/java/org/opensearch/dataprepper/plugins/source/dynamodb/stream/ShardAcknowledgementManager.java

```diff
@@ -0,0 +1,231 @@
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentLinkedQueue;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.TimeUnit;
+import java.util.function.Consumer;
+
+import org.opensearch.dataprepper.plugins.source.dynamodb.model.ShardCheckpointStatus;
+
+public class ShardAcknowledgementManager {
+    private static final Logger LOG = LoggerFactory.getLogger(ShardAcknowledgementManager.class);
+
+    private static final String NULL_SEQUENCE_NUMBER = "null";
+
+    private static final long WAIT_FOR_ACKNOWLEDGMENTS_TIMEOUT = 10L;
+
+    static final Duration CHECKPOINT_INTERVAL = Duration.ofSeconds(10);
+
+    private final DynamoDBSourceConfig dynamoDBSourceConfig;
+    private final Map<StreamPartition, ConcurrentLinkedQueue<ShardCheckpointStatus>> checkpoints = new ConcurrentHashMap<>();
+    private final ConcurrentHashMap<StreamPartition, ConcurrentHashMap<String, ShardCheckpointStatus>> ackStatus = new ConcurrentHashMap<>();
+
+    private final AcknowledgementSetManager acknowledgementSetManager;
+
+    private final EnhancedSourceCoordinator sourceCoordinator;
+
+    private final ExecutorService executorService;
+    private final List<StreamPartition> partitionsToRemove;
+    private boolean shutdownTriggered;
+
+    private Instant lastCheckpointTime;
+
+    public ShardAcknowledgementManager(final AcknowledgementSetManager acknowledgementSetManager,
+                                       final EnhancedSourceCoordinator sourceCoordinator,
+                                       final DynamoDBSourceConfig dynamoDBSourceConfig
+    ) {
+        this.acknowledgementSetManager = acknowledgementSetManager;
+        this.sourceCoordinator = sourceCoordinator;
+        this.dynamoDBSourceConfig = dynamoDBSourceConfig;
+        this.executorService = Executors.newSingleThreadExecutor(r -> {
+            Thread t = new Thread(r, "dynamodb-shard-ack-monitor");
+            t.setDaemon(true);
+            return t;
+        });
+        this.partitionsToRemove = Collections.synchronizedList(new ArrayList<>());
+        this.lastCheckpointTime = Instant.now();
+    }
+
+    public void init(final Consumer<StreamPartition> stopWorkerConsumer) {
+        executorService.submit(() -> monitorAcknowledgments(stopWorkerConsumer));
+    }
+
+    void monitorAcknowledgments(final Consumer<StreamPartition> stopWorkerConsumer) {
+        while (!Thread.currentThread().isInterrupted()) {
+            boolean exit = runMonitorAcknowledgmentLoop(stopWorkerConsumer);
+            if (exit) {
+                break;
+            }
+        }
+
+        LOG.info("Exiting acknowledgment manager");
+    }
+
+    boolean runMonitorAcknowledgmentLoop(final Consumer<StreamPartition> stopWorkerConsumer) {
+        removePartitions();
```

Comment:
Why do you remove all partitions up front?

Reply:
Yes this code basicallly handles failures (i.e. negative acknowledgment or ack timeout) and keeps us from continuing to check those once they fail. Could rename to `stopTrackingFailedPartitions` or something like that.

Reply:
Yea, let's give it a clearer name.